### DISCUSSION TITLE
Input-leap Wayland Mega Thread

### DISCUSSION THREAD

**Discussion by sithlord48** [input-leap/input-leap#1976] [Development of Input Leap]
A Place to put all the wayland things 
Related https://github.com/deskflow/deskflow/discussions/7499

# Desktop Info Working Environments

Notes on Specific desktops / portals 

## Working Desktops 
  - Kde Plasma 6.1+
  - GNOME 46+

## Portals That do not support Input-Capture 
You need to have a portal with Input-Capture support. (the one in use depends on your DE)
   - gtk (https://github.com/flatpak/xdg-desktop-portal-gtk/issues/494)
   - wlroots (https://github.com/emersion/xdg-desktop-portal-wlr/issues/278)
   - lxde (https://github.com/lxqt/xdg-desktop-portal-lxqt/issues/42)
   - elementary (https://github.com/elementary/portals/issues/117)
   - deepin (https://github.com/orgs/linuxdeepin/discussions/10659)
   - xapp (https://github.com/linuxmint/xdg-desktop-portal-xapp/issues/21)
   - hyperland(https://github.com/hyprwm/xdg-desktop-portal-hyprland/issues/259)

# Input-leap Specific Wayland Bugs
Place to list known bugs when running on wayland

# Known Upstream Issues
Bugs in upstream that affect us our use on wayland
-  Copy and paste doesn't work on Wayland (https://github.com/flatpak/xdg-desktop-portal/issues/1434)
- Permissions dialog shows every time (https://github.com/flatpak/xdg-desktop-portal/pull/1186)

**Comment by Esras** [comment]
Hi there!

Since the upstreaming of the Input Portal changes in KDE Plasma 6.1, I decided to test out Input Leap with my Fedora 40 Server and MacOS client setup. Spoilers: I did get it to work, but there were some extra steps and some pieces i want to share. This is a bit of a step-by-step of what I did and why, and then in a later section I have what is/"isn't" working.

# Preface
My primary driver is a Fedora 40 Workstation desktop. I utilize Wayland as my default environment now, via Plasma. I am running an Nvidia GPU (because even in places where that doesn't matter, it somehow matters). My client device is an M3 Pro MacBook.

I wanted to stop using a (poor) physical USB KVM since the MacBook wouldn't properly send reinitialize signals for USB devices.

The rpm packaged by the GH Action tries to run the server through XWayland, which didn't work for me, which led me to building it locally.

# Building
## Dependencies
For Fedora 40, the following dependencies are also necessary: `qt6-qtbase` `qt6-qtbase-devel` `qt6-qttools-devel`.

## CMake
There are two relevant CMake flags: `INPUTLEAP_BUILD_LIBEI` and `INPUTLEAP_BUILD_X11`. My first inclination was to build without X11, but that doesn't work right now. I added it as an extra line in the `clean_build.sh` build script:
```
...
B_CMAKE_FLAGS="-DINPUTLEAP_BUILD_LIBEI=ON"
B_BUILD_DIR ...
```

Running the build at this point succeeded. But, when running it, I get the following error in the logs:
```
[timestamp] FATAL: An error occurred: Missing libportal InputCapture portal support
	.../input-leap/src/lib/inputleap/App.cpp,115
```

## `libportal`
On Fedora, `libportal`'s version is `0.7.1-3.fc40`. Version 0.8.x released with support for the Input Capture portal, but that was only at the beginning of September. You can see that it's [not even in Rawhide yet](https://packages.fedoraproject.org/pkgs/libportal/libportal/) (as of writing). This [upstream issue](https://bugzilla.redhat.com/show_bug.cgi?id=2309405) is tracking that the update is available, for Red Hat. (Also of note, I'm not sure that's the _exact_ thing that was released that fixed later issues, but the newer version works and the older one didn't. Could be qt6, since I don't have `libportal-qt6` on my machine or available at all in the Fedora repos.)

So, the "clear" solution is to download and build it. I found [this](https://discussion.fedoraproject.org/t/input-leap-fedora-39-gnome-wayland/95856/9) thread where Benny Powers described their process of building input-leap from scratch. An additional dependency I needed was `vala` for `vapigen`, and the important part here is `meson _build`, `ninja -C _build`, and `cd _build && sudo meson install`.

Then, you can go back and re-run the build for input-leap.

# Running
So, success, right?

Actually, yes. At this point, running the server properly prompted me to allow something to have control (the portal support), and I was able to connect to the server from my MacBook.

# Bugs and Oddities
## Server (Fedora)
- When moving the mouse down on a client toward the bottom of the screen, KDE will treat it as a hovering event over the different windows in the docking bar or whatever it's called these days.
- "Shaking" the mouse around while on a client will have the KDE "mouse cursor growing" effect on the server.

## Client (MacOS)
- Scrolling with my mouse sends a single scroll "notch" and then ignores / eats any additional scroll events.
- Leaving the mouse idle for too long on the Client "resets" it to the host screens.
- If the mouse idles, it seems to briefly "freeze" before catching up to whatever movement is made. Looks similar to #300, including the part where if you down the `awdl0` interface, the performance improves. Also you only get that for a second or so because it'll immediately "restart."
- The freeze aspect seems to hold for the keyboard, as well, as there will be times where it takes a second or so for it to "catch" up to what you've typed. Sometimes this eats inputs.

**Comment by Shuna322** [comment]
Hyprland desktop portal has a promising PR
https://github.com/hyprwm/xdg-desktop-portal-hyprland/pull/268
But I think Hyprland still doesn't have a RemoteDesktop portal, is it a hard requirement for input-leap on Wayland?

**Reply by sithlord48** [reply]
Depends on your use IIRC the remote desktop is for clients and input capture for servers

**Comment by brianjmurrell** [comment]
https://github.com/input-leap/input-leap/releases/tag/v3.0.0 says:
> Added Wayland support. **Note that XWayland won't work properly and warning is printed.**

(emphasis is mine)

Can somebody comment (as if I am 5 years old since I don't use Wayland [because it's really still not ready to be a full replacement for X.org IMO] and have no idea what it's components do or mean) on what exactly this limitation with XWayland means to an end user?  How would an end-user experience the issues with XWayland?

**Reply by Esras** [reply]
If I recall correctly, it just doesn't meaningfully work without severe bugs that are limitations inherent to differences in X and Wayland's operating models.

---

Due to changes that were made in regards to the security model of Wayland vs. XOrg, there's a number of things that needed to get broken out into their own libraries. In particular, `libei` handles a lot of input events for Wayland-based compositors (the component that actually pulls together what the "screen" is).

Because there's so much software and there was a desire not to break it _all_, XWayland was created as an X Server running on the Wayland protocol. For _most_ cases, this means that an application built against X will work just fine and display all the things it needs to and take all the inputs it needs to.

(This line I'm a bit fuzzy on, so others feel free to correct me) Because of some of the isolation (a good thing) that Wayland implements, and the fact that X _didn't_ have those restrictions, there's a whole range of features that you can't just do "through the windowing manager." Some of those things look like hijacking user input. Like Input Leap.

So what that means is that if Input-Leap is running against XWayland (as in, it is the X version of the application), then the application will launch and show its window and act all happy, but the moment that you try to actually _use_ it, it's not going to be able to route the inputs correctly. There's a few different bugs that end up manifesting.

In my experience, I'd say Wayland is "there" for most "standard desktop" use cases, and if you're using an Nvidia card, they finally fixed some of the long-standing issues a few months ago. That being said, I have heard of some of the edge / corner cases that don't work in it, so that's not entirely a blanket statement.

**Reply by brianjmurrell** [reply]
Ah, right.  You have refreshed my memory that GTK (and therefore GNOME) can run directly on Wayland and eliminate the X layer in there.  Which is fine and dandy for GTK apps.  But if one has some GUI app that is still an Xlib app, of course it still needs an X display, enter XWayland.

So given that one is using GNOME, everything GNOME will work with Input Leap, but what happens with/to that Xlib app?  I suppose it only affects an Xlib app that is on an Input Leap client machine, yes?  But what is the result?  Keyboard (and mouse?) input on the Input Leap server machine don't make it to the Xlib app on the Input Leap client?

Or something else?

**Comment by benmordecai** [comment]
I am using the AUR packages input-leap-inputcapture, and libportal-inputcapture. However, these have just been deleted from the AUR and so are no longer maintained by anyone and I don't think you can even get them if you want them. 

I think I originally installed them with the release of Plasma 6.1 to the Arch repos. I am curious what is required to have a Arch, Plasma 6.2 server, but scared to try anything since I have a working system. Essentially, I'm worried I will be hosed for Wayland if I need to reinstall my Arch system for some reason. 

Is it working out of the box now?

**Reply by zhurai** [reply]
I've been just using the extra repository ones

```
[zhurai@home-pc01 ~]$ pacman -Q | grep -E "libportal|input-leap"
input-leap 3.0.2-1
input-leap-headless 3.0.2-1
libportal 0.8.1-1
libportal-gtk3 0.8.1-1
libportal-gtk4 0.8.1-1
libportal-qt5 0.8.1-1
libportal-qt6 0.8.1-1
```

Might be helpful to also check/note the current deps list https://aur.archlinux.org/packages/input-leap referencing the extra repo ones

I'd remove the old unmaintained ones you have installed and switch over to the extra repo ones.
